tests/test_fields.py

```diff
@@ -0,0 +1,73 @@
+import asyncio
+
+import attrs
+import pytest
+
+from web_poet import HttpResponse, ItemPage, field, item_from_fields
+
+
+@attrs.define
+class Item:
+    name: str
+    price: str
+
+
+@attrs.define
+class Page(ItemPage):
+    response: HttpResponse
+
+    @field
+    def name(self):  # noqa: D102
+        return self.response.css("title ::text").get()
+
+    @field
+    async def price(self):  # noqa: D102
+        await asyncio.sleep(0.01)
+        return "$123"
+
+    async def to_item(self):  # noqa: D102
+        return await item_from_fields(self, Item)
+
+
+@attrs.define
+class InvalidPage(ItemPage):
+    response: HttpResponse
+
+    @field
+    def name(self):  # noqa: D102
+        return self.response.css("title ::text").get()
+
+    @field
+    def unknown_attribute(self):  # noqa: D102
+        return "foo"
+
+    async def to_item(self):  # noqa: D102
+        return await item_from_fields(self, Item)
+
+
+@pytest.mark.asyncio
+async def test_fields():
+    response = HttpResponse(
+        "http://example.com",
+        b"""
+    <html>
+    <head><title>Hello!</title>
+    </html>
+    """,
+    )
+    page = Page(response=response)
+
+    assert page.name() == "Hello!"
```

Comment:
Note that fields in these PRs are preserved as extraction methods, they're not converted to properties. I don't have a super-strong opinon about it, but it’s done like this for a few reasons:

1. It makes it clear that computation could be expensive; IMO it’s much easier to miss it when accessing a property. With a method it's explicit that there is a potentially heavy computation which can happen. Less opportunity to forget about cache when it’s needed.
2. It allows to add optional arguments to these methods, if really needed. These arguments won’t be used in to_item, but could be used elsewhere.
3. Implementation is simpler (?), no need for concepts like “async property”.

Reply:
Point 3 sells it for me, as I have a feeling things can get complicated when mixing properties and both sync and async code.

I do wonder, thinking about 1, if [caching](https://github.com/scrapinghub/web-poet/pull/53/files#r927287307) wouldn’t be a good default (i.e. opt-out, rather than opt-in) feature of `@field`, as I imagine in most cases it will be desirable.

And as for 2, in real page objects I would not expect that, I would expect methods needing arguments to not be field-decorated, i.e. be some private methods that field-decorated methods call.

Which makes me think, if 3 were to be a non-issue (I have a *feeling* it can be an issue, but I do not know for a fact), maybe going for properties would not be such a bad idea, as it would discourage 2.

Reply:
About (1), there is a note in PEP8 (see https://peps.python.org/pep-0008/#designing-for-inheritance):

> Note 2: Avoid using properties for computationally expensive operations; the attribute notation makes the caller believe that access is (relatively) cheap.

Reply:
Having cache opt-out, not opt-in, sounds good.

Reply:
I tried to add cache by default to field, but I'm not sure if that's good or not. 

It seems it won't hurt much to have such cache. But, at the same time, it seems it's most useful to cache things which are not fields. For example, there could be an API call made with additional requests, and multiple attributes might be extracted from this call. User should create a method (or a property?) to do this call, and then use it in attributes:

```py

class MyPage(ItemPage):
    response: HttpResponse
    http: HttpClient

    @i_want_to_be_cached
    async def api_result(self):
        api_url = self.response.css("...").get()
        return self.http.get(api_url).json()

    @field
    def price(self):
        res = await self.api_result()
        return res['price']   

    @field
    def name(self):
        res = await self.api_result()
        return res['name']   
```

It's not that important to cache name and price attributes, but caching api_result is. Again, having cache for attributes probably won't hurt, but that doesn't seem to be the main concern.

Reply:
Actually, cache is more tricky for methods than for properties. I'll try to be careful with the implementation. For properties there is cached_property available in stdlib since Python 3.8, but for methods there is only lru_cache, which doesn't work properly for methods - `self` argument is used as a key in cache, and it prevents deallocation of instances.

Reply:
Maybe opt-out caching is not that good of an idea. I am OK with merging without any caching. Giving users freedom to choose their caching approach may work best. And implicit caching could be considered too magical, possibly confusing.

Reply:
Caching of method resuts is surprisingly tricky: built-in lru_cache doesn't work properly on methods, and workarounds which use weakrefs to `self` don't work either for Page Object methods, as they require POs to be hashabe, while they're not always hashable - e.g. if they have HttpResponse dependency, then they're not, because headers are not hashable.

I'd like to get some solution for this before merging this PR. The problem is solved more easily for properties.

Reply:
I've reverted field to not use cache by default, added `cached_method` decorator, and `field(cached=True)` shortcut. Todo: add documentation. 

Opt-out doesn't seem to be a good idea - it seems for most fields it's not going to be useful, and we may be increasing memory usage for no good reason.